designs/2025-rule-tester-assertion-options/README.md

```diff
@@ -0,0 +1,365 @@
+- Repo: https://github.com/eslint/eslint
+- Start Date: 2025-07-10
+- RFC PR: https://github.com/eslint/rfcs/pull/137
+- Authors: ST-DDT
+
+# Rule Tester Assertions Options
+
+## Summary
+
+Add options that control which assertions are required for each `RuleTester` test case.
+
+## Motivation
+
+In most eslint(-plugin)s' rules the error assertions are different from each other.
+Adding options that could be set/shared when configuring the rule tester would ensure a common base level of assertions is met throughout the (plugin's) project.
+The options could be defined on two levels. On `RuleTester`'s `constructor` effectively impacting all tests, or on the test `run` method itself, so only that set is affected.
+
+## Detailed Design
+
+```ts
+ruleTester.run("rule-name", rule, {
+  assertionOptions?: {
+    /**
+     * Require message assertion for each invalid test case.
+     * If `true`, each error must extend `string | RegExp | { message: string | RegExp } | { messageId: string }`.
+     * If `'message'`, each error must extend `string | RegExp | { message: string | RegExp }`.
+     * If `'messageId'`, each error must extend `{ messageId: string }`.
+     *
+     * @default false
+     */
+    requireMessage?: boolean | 'message' | 'messageId';
+    /**
+     * Require full location assertions for each invalid test case.
+     * If `true`, each error must extend `{ line: number, column: number, endLine?: number | undefined, endColumn?: number | undefined }`.
+     * `endLine` or `endColumn` may be absent or undefined, if the observed error does not contain these properties.
+     *
+     * @default false
+     */
+    requireLocation?: boolean;
+  },
```

Comment:
Another helpful check would be `requireData` which requires the `data` error property when a `messageId` is used and the referenced message has placeholders (`{{ placholder }}`).

Reply:
Would this imply `requireMessage: 'messageId'`?

Reply:
This option (`requireData`) should only be active when the error has a `messageId` as `data` is useless when `message` is specified (you already provide the message with every placeholder filled in).
The validator only has to check whether `data` is specified as there is already an error when `data` is specified and placeholders are missing.
Coupling this with the `requireMessage: "messageId"` would be unintuitive as that reads as only the "messageId" is required.

Reply:
I agree a `requireData` assertion could be helpful. As this would a separate option, it could be also discussed in a follow-up RFC.